voice_assistant_pkg/voice_assistant_pkg/voice_assistant.py

```diff
@@ -0,0 +1,246 @@
+key = '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'
+
+# openAI set-up
+openai.api_key = key
+messages  = [{"role": "system", "content": f"{personality}"}]
+
+# pyttsx3 setup: text to speech voice generator
+engine = pyttsx3.init()
+voices = engine.getProperty('voices')
+engine.setProperty('voice', voices[11].id) # 0 for male, 1 for female
+# for i, voice in enumerate(voices):
+#     if voice.id == 'english':
+#         print(i, voice)
+
+# speech recognition set-up
+r = sr.Recognizer()
+# print(sr.Microphone.list_microphone_names())
+mic = sr.Microphone(device_index=12) #11 or 12: self microphone, 13: snowball
+r.dynamic_energy_threshold=False
+r.energy_threshold = 200
+
+def get_input():
+    with mic as source:
+        print("\nAmbient noise adjust...")
+        r.adjust_for_ambient_noise(source)
+        print("\nListening...")
+        audio = r.listen(source)
+        try:
+            if usewhisper:
+                user_input = whisper(audio)
+                print('From whisper: ', user_input)
+            else:
+                user_input = r.recognize_google(audio)
+                print('Transcribed from google: ', user_input)
+            return user_input
+        except:
+            return ''
+
+
+def whisper(audio):
+    with open('speech.wav','wb') as f:
+        f.write(audio.get_wav_data())
+    speech = open('speech.wav', 'rb')
+    wcompletion = openai.Audio.transcribe(
+        model = "whisper-1",
+        file=speech
+    )
+    user_input = wcompletion['text']
+    return user_input
+
+
+def save_conversation(save_foldername):
+    '''
+    Checks the folder for previous conversations and will get the next suffix that has not been used yet.  
+    It returns suffix number
+
+    Args:
+        save_foldername (str) : Takes in the path to save the conversation to.
+    '''
+    
+    os.makedirs(save_foldername, exist_ok=True)
+
+    base_filename = 'conversation'
+    suffix = 0
+    filename = os.path.join(save_foldername, f'{base_filename}_{suffix}.txt')
+
+    while os.path.exists(filename):
+        suffix += 1
+        filename = os.path.join(save_foldername, f'{base_filename}_{suffix}.txt')
+
+    with open(filename, 'w') as file:
+        json.dump(messages, file, indent=4)
+
+    return suffix
+
+def save_inprogress(suffix, save_foldername):
+    '''
+    Uses the suffix number returned from save_conversation to continually update the 
+    file for this instance of execution.  This is so that you can save the conversation 
+    as you go so if it crashes, you don't lose to conversation.  Shouldn't be called
+    from outside of the class.
+
+    Args:
+        suffix  :  Takes suffix count from save_conversation()
+    '''
+    os.makedirs(save_foldername, exist_ok=True)
+    base_filename = 'conversation'
+    filename = os.path.join(save_foldername, f'{base_filename}_{suffix}.txt')
+
+    with open(filename, 'w') as file:
+        json.dump(messages, file, indent=4)
+
+class RunServoClient(Node):
```

Comment:
wait, is this for the servo motors? in the voice file?

Reply:
+1, this shouldn't be here. Voice recognition should be completely independent from actuation

Reply:
I agree. I did that on purpose so we can discuss what the best way to organize this is (since I have limited ros experience). Should I keep RunServoClient class in a separate package called RunServosClient and then import this class in voice_assistant.py? Organize the voice_assistant spaghetti code into a class that subscribes to human detection topic and instantiates RunServoClient node and calls its send_request member function upon receiving a voice request from the user. Or is there another more suitable/better way?

Reply:
Will any other node want to operate the servos besides the voice_assistant?